kernel/cls/cls_macros/src/lib.rs

```diff
@@ -87,22 +148,114 @@ pub fn cpu_local(args: TokenStream, input: TokenStream) -> TokenStream {
         Span::call_site().into(),
     );
 
-    let ((x64_asm_width, x64_reg_class), (aarch64_reg_modifier, aarch64_instr_width)) =
-        match ty.to_token_stream().to_string().as_ref() {
-            "u8" => (("byte", quote! { reg_byte }), (":w", "b")),
-            "u16" => (("word", quote! { reg }), (":w", "w")),
-            "u32" => (("dword", quote! { reg }), (":w", "")),
-            "u64" => (("qword", quote! { reg }), ("", "")),
-            _ => {
-                Diagnostic::spanned(ty.span().unwrap(), Level::Error, "invalid type")
-                    .help("CPU locals only support these types: `u8`, `u16`, `u32`, `u64`")
-                    .emit();
-                return TokenStream::new();
+    let ref_expr = quote! {
+        {
+            let mut ptr: usize;
+            #[cfg(target_arch = "x86_64")]
+            {
+                unsafe {
+                    ::core::arch::asm!(
+                        "rdgsbase {}",
+                        out(reg) ptr,
+                        options(nomem, preserves_flags, nostack),
+                    )
+                };
+            }
+            #[cfg(target_arch = "aarch64")]
+            {
+                unsafe {
+                    ::core::arch::asm!(
+                        "mrs {}, tpidr_el1",
+                        out(reg) ptr,
+                        options(nomem, preserves_flags, nostack),
+                    )
+                };
+            }
+            ptr += #offset;
+            unsafe { &mut*(ptr as *mut #ty) }
```

Comment:
> rather than writing our own asm here, let's just use existing Rust-level crates to access GSBASE (x86_64) or TPIDR_EL1 (cortex_a).

The only reason I was hesitant to do that because it means I would have to reexport them in `cls::__private`, but it's not that big of a deal.

> At least on x86 it's generally faster to read from %gs:0 than the GSBASE register

So I'm assuming you're referring to [this optimisation](https://github.com/rust-osdev/x86_64/pull/257). The reason using a self ptr is faster, is because first loading the value of the pointer in to a register means Rust can reason about and optimise the memory access. We do the same thing except using the `rdgsbase` instruction, which is specifically designed to read the gs base register efficiently. So, I don't think this applies to our use case.

Reply:
I was just asking about doing `%gs:0` vs `rdgsbase` alone, I realize they both allow the compiler to perform the same reasoning. It's likely not a big deal, I honestly did not know about `rsgsbase` before, so if that's the preferred efficient way to do it then we can skip the CLS self pointer entirely (unless it ends up being required for a TLS-based CLS approach, but we still don't have to use it)

Reply:
I doubt that `%gs:0` is faster than even `msr` access, because then other operating systems like Linux would be using that method.

Reply:
Can't argue with that, yeah

Reply:
So option 1 is good then as long as it compiles down to the same thing as our inline asm. Just trying to reduce inline asm or dedup it as much as possible, y'know